rustls/src/verify.rs

```diff
@@ -92,6 +95,57 @@ impl ClientCertVerified {
 /// signatures made by certificates.
 #[allow(unreachable_pub)]
 pub trait ServerCertVerifier: Send + Sync {
+    /// Verify the end-entity certificate and a signature by allegedly signed by it.
+    ///
+    /// The default implementation of this method just calls [`Self::verify_server_cert`]
+    /// and [`Self::verify_tls12_signature()`] or [`Self::verify_tls13_signature()`]
+    /// depending on the TLS version (in `input.version()`) for the given handshake.
+    ///
+    /// Verify the end-entity certificate `input.end_entity` is valid for the
+    /// hostname `input.server_name` and chains to at least one trust anchor.
+    ///
+    /// `input.intermediates` contains the intermediate certificates the client sent
+    /// along with the end-entity certificate; it is in the same order that the
+    /// peer sent them and may be empty.
+    ///
+    /// `input.scts()` contains the Signed Certificate Timestamps (SCTs) the server
+    /// sent with the certificate, if any.
+    ///
+    /// The `input.server_verify_message()` is not hashed, and needs hashing during
+    /// the verification. The signature and algorithm can be found as part of
+    /// `input.server_signature`.  `input.end_entity` contains the public key to use.
+    ///
+    /// Note that none of the certificates have been parsed yet, so it is the responsibility of
+    /// the implementor to handle invalid data. It is recommended that the implementor returns
+    /// [`Error::InvalidCertificateEncoding`] when these cases are encountered.
+    fn verify(
+        &self,
+        input: ServerVerifyInput,
+    ) -> Result<(ServerCertVerified, HandshakeSignatureValid), Error> {
```

Comment:
I disagree that returning a 2-tuple is particularly awkward (in Rust). I guess we could add a grouped marker type, but that feels like a somewhat orthogonal change.

Reply:
Here's why I think it's awkward in Rust: as an API reader, when I see a Result that returns a tuple, I assume it's because the function has multiple outputs. However this function doesn't have multiple outputs: the tuple is equivalent to the unit type, in that there is only one possible value for it. As a user, my next question is: what is the functionality that justifies having two distinct marker types in the tuple? Am I expected to split them up and use them separately at some point? Or is the library going to split them up and use them separately? Neither is the case; all the library cares about is that an assertion of validity for the certificate and the handshake occurred. So the API should express that in the simplest way possible.

I agree that for the internal code, a grouped marker type is an orthogonal change, but I think for the public API it makes sense to express the simplest possible marker type. One way to avoid plumbing through additional changes would be to go the other way: have a `ServerCertAndHandshakeVerified` that can be decomposed into a `ServerCertVerified` and `HandshakeSignatureValid`.